src/cats/CatSlideshow.js

```diff
@@ -0,0 +1,219 @@
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH = 8;
+const MAX_PREFETCH = 2;
+
+function fetchPageFromCatsApi(pageIndex, selectedBreedID, fetchPage) {
+  const { url, options } = getCatsApiFetchParams("/images/search", {
+    breed_id: selectedBreedID === "all" ? "" : selectedBreedID,
+    limit: PAGE_SIZE,
+    order: "ASC",
+    page: pageIndex,
+  });
+
+  const getPageData = async (response) => {
+    const json = await response.json();
+    return json.map((item) => ({ id: item.id, url: item.url }));
+  };
+
+  const getMetadata = (response) => {
+    return Object.fromEntries(
+      [...response.headers.entries()].filter(
+        (entry) => entry[0] === "pagination-count"
+      )
+    );
+  };
+
+  fetchPage(url, options, pageIndex, selectedBreedID, getPageData, getMetadata);
+}
+
+function getPageIndex(index) {
+  return Math.floor(index / PAGE_SIZE);
+}
+
+function getImageURL(pages, index) {
+  const pageIndex = getPageIndex(index);
+  const offset = index % PAGE_SIZE;
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+function initCatSlideshowState() {
+  return {
+    direction: Direction.Next,
+    index: 0,
+    maxIndex: null,
+    visibleIndex: null,
+    readyMap: {},
+  };
+}
+
+function catSlideshowReducer(state, action) {
+  //console.log("%ccat slideshow: " + action.type, "color: red", action, state);
+
+  function changeIndex(state, increment) {
+    const newIndex = Math.min(
+      Math.max(0, state.index + increment),
+      state.maxIndex
+    );
+    const isReady = state.indexReadyMap[newIndex] === "ready";
+
+    return {
+      ...state,
+      direction: increment > 0 ? Direction.Next : Direction.Previous,
+      index: newIndex,
+      visibleIndex: isReady ? newIndex : state.visibleIndex,
+    };
+  }
+
+  switch (action.type) {
+    case "change-breed":
+      return initCatSlideshowState();
+    case "update-max-index":
+      return { ...state, maxIndex: action.maxIndex };
+    case "decrement-index":
+      return changeIndex(state, -1);
+    case "increment-index": {
+      return changeIndex(state, 1);
+    }
+    case "prefetch-status": {
+      const newState = {
+        ...state,
+        indexReadyMap: { ...state.indexReadyMap, [action.index]: "ready" },
+      };
+
+      if (action.index === state.index) {
+        newState.visibleIndex = action.index;
+      }
+
+      return newState;
+    }
+    default:
+      throw new Error(`Unhandled action type: ${action.type}`);
+  }
+}
+
+function CatSlideshow({ selectedBreedID }) {
+  const prevSelectedBreedIDRef = useRef(selectedBreedID);
+  const prefetchMapRef = useRef(new Map());
+  const { pages, metadata, fetchPage, resetPages } = usePaginatedFetch(
+    PAGE_SIZE,
+    "loading"
+  );
+
+  const [state, dispatch] = useReducer(
+    catSlideshowReducer,
+    undefined,
+    initCatSlideshowState
+  );
+
+  const { direction, index, visibleIndex } = state;
+
+  useEffect(() => {
+    if (selectedBreedID !== prevSelectedBreedIDRef.current) {
```

Comment:
Since selectedBreedID is coming through via props, I don't need a useEffect at all. I can just delete line 119 and 126. Do I have that right?

Reply:
I'll take another look at how you are doing it now, but in general if you have to check for when props are changing, it is possible to instead react to the signal that caused the prop to change in the first place. (e.g. instead of having a prop named `scrollTop` in which you look for changes, you could call an imperative method in the scroll event itself)